src/util/piped_process.cpp

```diff
@@ -0,0 +1,388 @@
+  // Create the child process
+  STARTUPINFOW start_info;
+  BOOL success = FALSE;
+  ZeroMemory(&proc_info, sizeof(PROCESS_INFORMATION));
+  ZeroMemory(&start_info, sizeof(STARTUPINFOW));
+  start_info.cb = sizeof(STARTUPINFOW);
+  start_info.hStdError = child_std_OUT_Wr;
+  start_info.hStdOutput = child_std_OUT_Wr;
+  start_info.hStdInput = child_std_IN_Rd;
+  start_info.dwFlags |= STARTF_USESTDHANDLES;
+  // Unpack the command into a single string for Windows API
+  std::wstring cmdline = widen(commandvec[0]);
+  for(int i = 1; i < commandvec.size(); i++)
+  {
+    cmdline.append(L" ");
+    cmdline.append(quote_windows_arg(widen(commandvec[i])));
+  }
+  // Note that we do NOT free this since it becomes part of the child
+  // and causes heap corruption in Windows if we free!
+  WCHAR *szCmdline =
+    reinterpret_cast<WCHAR *>(malloc(sizeof(WCHAR) * cmdline.size()));
+  wcscpy(szCmdline, cmdline.c_str());
+  success = CreateProcessW(
+    NULL,        // application name, we only use the command below
+    szCmdline,   // command line
+    NULL,        // process security attributes
+    NULL,        // primary thread security attributes
+    TRUE,        // handles are inherited
+    0,           // creation flags
+    NULL,        // use parent's environment
+    NULL,        // use parent's current directory
+    &start_info, // STARTUPINFO pointer
+    &proc_info); // receives PROCESS_INFORMATION
+  // Close handles to the stdin and stdout pipes no longer needed by the
+  // child process. If they are not explicitly closed, there is no way to
+  // recognize that the child process has ended (but maybe we don't care).
+  CloseHandle(child_std_OUT_Wr);
+  CloseHandle(child_std_IN_Rd);
+#else
+  if(pipe(pipe_input) == -1)
+  {
+    throw std::runtime_error("Input pipe creation failed");
+  }
+  if(pipe(pipe_output) == -1)
+  {
+    throw std::runtime_error("Output pipe creation failed");
+  }
+  if(fcntl(pipe_output[0], F_SETFL, O_NONBLOCK) < 0)
+  {
+    throw std::runtime_error("Setting pipe non-blocking failed");
+  }
+  // Create a new process for the child that will execute the
+  // command and receive information via pipes.
+  pid = fork();
+  if(pid == 0)
+  {
+    // child process here
+    // Close pipes that will be used by the parent so we do
+    // not have our own copies and conflicts.
+    close(pipe_input[1]);
+    close(pipe_output[0]);
+    // Duplicate pipes so we have the ones we need.
+    dup2(pipe_input[0], STDIN_FILENO);
+    dup2(pipe_output[1], STDOUT_FILENO);
+    dup2(pipe_output[1], STDERR_FILENO);
+    // Create a char** for the arguments (all the contents of commandvec
+    // except the first element, i.e. the command itself).
+    char **args =
+      reinterpret_cast<char **>(malloc((commandvec.size()) * sizeof(char *)));
+    // Add all the arguments to the args array of char *.
+    unsigned long i = 0;
+    while(i < commandvec.size())
+    {
+      args[i] = strdup(commandvec[i].c_str());
+      i++;
```

Comment:
No need for the `strdup`, and the subsequent `free`.

Reply:
The child process wants to be able to modify the arguments (i.e. non `const`) while the `commandvec`  with `c_str()`can only provide `const` instances.

Reply:
This memory is not shared with the child; the child gets a copy. `const` is fine.

Reply:
Fixed. :)

Reply:
Please add the `const` to the type of `args`. There is then no need for the `const_cast`.